goopenssl.c

```diff
@@ -57,11 +57,24 @@ go_openssl_fips_enabled(void* handle)
         return FIPS_mode();
 
     // For OpenSSL 3.x.
-    int (*EVP_default_properties_is_fips_enabled)(void*) = (int (*)(void*))dlsym(handle, "EVP_default_properties_is_fips_enabled"); 
-    if (EVP_default_properties_is_fips_enabled != NULL && EVP_default_properties_is_fips_enabled(NULL) == 1)
-            return 1;
+    int (*EVP_default_properties_is_fips_enabled)(void*) = (int (*)(void*))dlsym(handle, "EVP_default_properties_is_fips_enabled");
+    void *(*EVP_MD_fetch)(void*, const char*, const char*) = (void* (*)(void*, const char*, const char*))dlsym(handle, "EVP_MD_fetch");
+    void (*EVP_MD_free)(void*) = (void (*)(void*))dlsym(handle, "EVP_MD_free");
 
-    return 0;
+    if (EVP_default_properties_is_fips_enabled == NULL || EVP_MD_fetch == NULL || EVP_MD_free == NULL) {
+        // Shouldn't happen, but if it does, we can't determine if FIPS mode is enabled.
+        return 0;
```

Comment:
A false negative for FIPS mode does seem nearly harmless, but it also seems like it would be pretty simple to return and handle `-1` or `2` or something just to be sure.

Reply:
The only thing with the current `openssl.CheckVersion` API is to log the error or panic, as it doesn't return any error value. I wouldn't say it is a false negative, if those functions are not available then we won't be able to use the shared library properly, even less in FIPS mode.

Reply:
The sequence of events I'm thinking about is that the user called `CheckVersion(version string) (exists, fips bool)` on a "bad" version of OpenSSL that doesn't define one of these functions, and then `CheckVersion` returns `true, false`. This implies that it exists and *isn't* in FIPS mode (given the current docs), but technically we just can't tell what mode it's in.

I guess the more realistic oddity is the `exists = true`. If the caller of `CheckVersion` is trying to decide which version to load based on this info, they might try to use the broken version that we already know we can't use properly. Maybe it would be better to return `false, false`?

In practice I don't think this is likely to ever matter, the code just makes this edge case relatively easy to spot. 😄

Reply:
> I guess the more realistic oddity is the exists = true.

Good point. In fact, the current implementation returns `exists=true` if the shared library loaded is not OpenSSL, which is weird. I'll update the code with your recommendations. Thanks!